// nearbytes-storage/dist/storage/pathRecord.js
export function normalizeStoragePath(path) {
    return path.replace(/^\/+/, '').replace(/\/+/g, '/').replace(/\/$/, '').trim();
}
export function normalizeStorageDirectoryPath(path) {
    const normalized = normalizeStoragePath(path);
    return normalized === '' ? '' : `${normalized}/`;
}
export function createInMemoryPathRecordStore(initial) {
    return {
        files: initial?.files ?? new Map(),
        directories: initial?.directories ?? new Set(),
    };
}
export function createInMemoryPathRecordStoreAdapter(store) {
    return {
        async putRecord(path, data) {
            store.files.set(path, {
                path,
                data: new Uint8Array(data),
                updatedAt: Date.now(),
            });
        },
        async getRecord(path) {
            const record = store.files.get(path);
            if (!record) {
                return null;
            }
            return {
                path: record.path,
                data: new Uint8Array(record.data),
                updatedAt: record.updatedAt,
            };
        },
        async deleteRecord(path) {
            store.files.delete(path);
        },
        async putDirectory(path) {
            store.directories.add(path);
        },
        async hasDirectory(path) {
            return store.directories.has(path);
        },
        async listStoredPaths() {
            return Array.from(store.files.keys());
        },
    };
}
export class PathRecordStorageBackend {
    constructor(store) {
        this.store = store;
    }
    async writeFile(path, data) {
        await this.store.putRecord(normalizeStoragePath(path), data);
    }
    async readFile(path) {
        const record = await this.store.getRecord(normalizeStoragePath(path));
        if (!record) {
            throw new Error(`File not found: ${path}`);
        }
        return new Uint8Array(record.data);
    }
    async listFiles(directory) {
        const prefix = normalizeStorageDirectoryPath(directory);
        const paths = await this.store.listStoredPaths();
        const files = new Set();
        for (const path of paths) {
            if (prefix !== '' && !path.startsWith(prefix)) {
                continue;
            }
            const remainder = prefix === '' ? path : path.slice(prefix.length);
            if (remainder === '' || remainder.includes('/')) {
                continue;
            }
            files.add(remainder);
        }
        return Array.from(files).sort((left, right) => left.localeCompare(right));
    }
    async createDirectory(path) {
        await this.store.putDirectory(normalizeStoragePath(path));
    }
    async exists(path) {
        const normalized = normalizeStoragePath(path);
        if (normalized === '') {
            return true;
        }
        if (await this.store.getRecord(normalized)) {
            return true;
        }
        if (await this.store.hasDirectory(normalized)) {
            return true;
        }
        const prefix = `${normalized}/`;
        const paths = await this.store.listStoredPaths();
        return paths.some((entry) => entry.startsWith(prefix));
    }
    async deleteFile(path) {
        await this.store.deleteRecord(normalizeStoragePath(path));
    }
}
